doc/xfreerdp.1
.TH xfreerdp 1 "May 2010"
.SH NAME
.I xfreerdp
\- X11 front-end for FreeRDP
.SH SYNOPSIS
.B xfreerdp [options] server[:port] [[options] server[:port] ...]
.br
.SH DESCRIPTION
.I xfreerdp
is a front-end for various FreeRDP components, which implement a client of the
Remote Desktop Protocol (RDP), used in a number of Microsoft products including
Windows NT Terminal Server, Windows 2000 Server, Windows XP, Windows 2003
Server, Windows Vista, Windows 2008 Server and Windows 7, but also implemented
by xrdp and VirtualBox.

xfreerdp can connect to multiple servers if more than one server
addresses are specified, and each session runs in a separate window.
.SH OPTIONS
.TP
.BR "-u <username>"
Username for authentication on the server.
.TP
.BR "-d <domain>"
Domain for authentication.
.TP
.BR "-p <password>"
The password to authenticate with.  Note that this may have no effect if
"Always prompt for password" is enabled on the server.  WARNING: if you specify
a password on the command line it may be visible to other users when they use
tools like ps.
.TP
.BR "-s <shell>"
Startup shell for the session - starts a specific application instead of Explorer.
.TP
.BR "-c <directory>"
The initial working directory for the session.  Often used in combination with -s
to set up a fixed login environment.
.TP
.BR "-n <hostname>"
Client hostname reported to the server.  Normally xfreerdp automatically
obtains the hostname of the client.
.TP
.BR "-t <port>"
Server port number. Default is the standard RDP port 3389. Can also be
specified as ":<port>" after the server hostname.
.TP
.BR "-g <geometry>"
Desktop geometry (WxH). 
.TP
.BR "-a <bpp>"
Sets the color depth for the connection (8, 15, 16, 24 or 32).
More than 8 bpp are only supported when connecting to Windows XP
(up to 24 bpp) or newer.  Note that the color depth may also be
limited by the server configuration. The default value is the depth 
of the root window. 
.TP
.BR "-z"
Enable compression inside the encrypted RDP datastream.
.TP
.BR "-f"
Full screen mode. Causes the window to fully cover the current screen.
This mode can be toggled at any time using Ctrl-Alt-Enter
.TP
.BR "-D"
Hide window decorations.
.TP
.BR "-x <experience>"
Changes default bandwidth performance behaviour for RDP5. By default only
theming is enabled, and all other options are disabled (corresponding
to modem (56 Kbps)). Setting experience to b[roadband] enables menu
animations and full window dragging. Setting experience to l[an] will
also enable the desktop wallpaper. Setting experience to m[odem]
disables all (including themes). Experience can also be a hexidecimal
number containing the flags.
.TP
.BR "-0"
Attach to the admin console of the server. This is the default if no
server license is available.
.TP
.BR "-o"
Play audio on the server side.
.TP
.BR "-k"
Force particular keyboard layout ID.
.TP
.BR "--kbd-list"
List all available keyboard layout IDs and exit.
.TP
.BR "--no-tls"
Disable TLS encryption.
.TP
.BR "--no-osb"
Disable off screen bitmaps.
.TP
.BR "-h"
Print help message.
.TP
.BR "--version"
Print version information.
.TP
.BR "--plugin <pluginname> [--data <plugindata> --]"
Load a Virtual Channel plugin module, and optionally pass initial parameters
to it. <pluginname> can be the virtual channel name, or a path specifying the
plugin module. If the virtual channel name was passed, xfreerdp will try to
locate the plugin module in its plugin path ${prefix}/lib/freerdp.

Some plugins support retrieving initial parameters. Add --data argument after
the <pluginname> in order to pass such parameters, and end it with --.

The --plugin argument can be repeated more than once if there are multiple
plugins to load.

Following plugins are currently supported:
.TP
.BR "--plugin cliprdr"
Synchronize client and server clipboard data. Plain text, Unicode text, HTML
content and images are supported.
.TP
.BR "--plugin rdpsnd [--data [<subplugin>] [<param>] ... --]"
Redirects sound generated on the server to the client. <subplugin> can be one
of the following:

.B
alsa[:<alsa_device>]
Redirect sound to ALSA device <alsa_device>. If :<alsa_device> is omitted,
sound will be redirected to 'default' device.

.B
pulse[:<pulse_device>]
Redirect sound to PulseAudio device <pulse_device> on the default PA server.
If :<pulse_device> is omitted, default device is used.

<param> can be a combination of the following:

.B
format:n
Force to use specific audio format. Possible values are 1(PCM), 6(A-LAW),
7(MU-LAW) and 17(Microsoft IMA ADPCM). If omitted, the server will usually
choose the format with best compression.

.B
rate:n
Force to use specific sample rate. Possible values are 8000, 11025, 22050 and
44100. If omitted, the server will choose one that is best suited for the
network speed and latency.

.B
channel:n
Force to use specific number of channels. Possible values are 1(mono) and
2(stereo).

.TP
.BR "--plugin rdpdr --data <subplugin> [<subplugin> ...] --"
Redirects file system devices on your client to the server. <subplugin> can be
one of more of the following:

.B
disk:<sharename>:<path>
Redirect <path> to the server as shared folder \\\\tsclient\\<sharename>.

.B
printer[:<printername>[:<driver>]]
Redirect printers to the server. If both <printername> and <driver> are
omitted, the printer sub-plugin will automatically redirect all CUPS printers
using the default PostScript driver "MS Publisher Imagesetter".

.B
serial:<sharename>:<device>
Redirect serial <device> to the server. It will be referenced by <sharedname>.

.B
parallel:<sharedname>:<lptdevice>
Redirects parallel device <lptdevice> on your client to the server, where it
will be referenced by <sharedname>. Bidirectional/Read support requires
Windows XP or newer.

.TP
.BR "--plugin drdynvc --data <subplugin> [<subplugin> ...] --"
Load the Dynamic Virtual Channel. <subplugin> can be the DVC sub-plugin name,
or a path specifying the sub-plugin module. If the DVC sub-plugin name was
passed, xfreerdp will try to locate the sub-plugin in its plugin path
${prefix}/lib/freerdp.

Following DVC sub-plugins are currently supported:

.B
audin
Redirect audio recording device to the server. This is an RDP 7.0 feature
available in Windows 7, Windows 2008 and Windows 2008 R2. Note that Windows 7
Enterprise edition and Windows 2008 Server has audio redirection disable by
default.

.PP
.SH LINKS
Main website of FreeRDP
.br
\fIhttp://www.freerdp.com/
.LP
.PP


ChangeLog
2010-11-13  Version 0.8.2
   * libfreerdp:
    - fix buffer overflow when local timezone doesn't fit in 32 bytes
    - Fix data corruption after TLS license package
    - free TLS resources to avoid memory leaks
   * mcs:
    - Fix connecting to VirtualBox with extra channels
   * wfreerdp: 
    - align command line options with xfreerdp
    - Fixes for compilation with VS2008
   * rdpsnd plugin: 
    - ensure the ALSA device is closed when freeing resources
   * rdpdr plugin:
    - fix various memory leaks
    - fix a segfault in multi-session environment

2010-10-25  Version 0.8.1
   * cunit tests:
    - added --with-cunit compile time option (disabled by default)
    - cunit tests now included in distribution tarball 
   * wfreerdp:
    - removed unused ASN.1 MCS source files that would affect wfreerdp
    - fixed various problems that would prevent wfreerdp from compiling

2010-10-24  Version 0.8.0
   * xfreerdp: 
    - show less debug info unless in debug mode
    - more useful keyboard debugging
    - new command-line-option: --no-osb: disable off screen bitmaps
    - Host-names now appear in window-title
    - new command-line-option: --version
   * rdpdr plugin:
    - some fixes
    - parallel port redirection
    - serial port redirection
   * drdynvc (Dynamic Virtual Channel) plugin:
    - started work
    - added audin (Redirect audio recording device to the server) subplugin (EXPERIMENTAL)
   * Added working TLS encryption support:
    - enable TLS by default and added option --no-tls (TLS is currently only available with OpenSSL as crypto backend)
   * Crypto restructuring
    - added partial support for PolarSSL as crypto backend
    - added support for NSS as crypto backend
    - added support for GnuTLS as crypto backend
   * Fixes for Load Balancing and authentication when redirecting

2010-08-18  Version 0.7.4
   * Maintenance:
    - Brazilian ABNT keyboard issue
    - libfreerdpkbd.c zombies
    - manpage fixes
    - VirtualBox connectivity

2010-07-23  Version 0.7.3
   * Maintenance:
    - Add freerdp_ prefix to get_wstr/set_wstr to avoid function name conflict
    - Corrected neo keyboard layout id

2010-07-04  Version 0.7.2
   * xfreerdp: 
    - --help as alias for -h
    - report invalid options to the user 
    - assuming no Host-Name can start with "-"
    - added German neo layout to list of keyboard layout variants
   * Maintenance:
    - Fixes in crypto
    - Works towards DirectFB
   
2010-06-19  Version 0.7.1
   * xfreerdp: 
    - added -k option to force particular keyboard layout ID
    - added --kbd-list option to list all keyboard layout IDs
    - added -h option for help
    - changed -plugin and -data argument to --plugin and --data
   * Maintenance:
    - Fixes to build system

2010-06-11  Version 0.7.0
   * Software Architecture:
    - Full multi-session support
    - Created a library separate from the UI, making it easier to implement different UI's
    - New keyboard input system based on the XKB database, providing better support for international keyboards
    - Implemented Microsoft's virtual channel plugin API and re-implemented features as plugins (libfreerdpchanman)
    - Created a "device manager" for various devices redirected over the device redirection plugin
   * Refactoring:
    - X11 UI completely rewritten from scratch
    - Disk and printer redirection have been rewritten in a much cleaner way
    - redirection has been re-implemented as a virtual channel plugin
    - Sound redirection has also been re-implemented as a virtual channel plugin
    - A *LOT* of added comments in the code with regards to the protocol documentation
    - Cleanup and increased compliance with the documentation released by MS after rdesktop was created
   * Improvements:
    - Support for cursors on Windows Server 2008 R2
    - Support for bitmap caching, which provides a good performance boost
    - Build system has been completely revamped to use newer tools such as libtool
    - Support for Windows Server 2008 Terminal Services Session Broker Load Balancing
    - It is now possible to compile FreeRDP without OpenSSL, using built-in code (more lightweight on embedded systems but only partially supported)
    - Real capability set processing (original implementation discarded capability sets almost entirely and sent blobs)
   * Work In Progress:
    - A DirectFB UI has been started
    - New Win32 port with working native UI, but it's not complete yet
    - Support for NSS as a replacement for OpenSSL is also being worked on
    - Network Level Authentication is half-way done (TLS works, but NTLM authentication is partially implemented)
   * Dropped or Broken Features:
    - seamlessrdp is *NOT* supported. RemoteApp support is eventually going to replace seamlessrdp
    - Serial port redirection is currently not supported, as it needs to be re-implemented as a plugin
    - Smartcard support also needs to be re-implemented as a plugin



X11/xfreerdp.c
/*
   Copyright (c) 2009-2010 Jay Sorg

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
   DEALINGS IN THE SOFTWARE.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <errno.h>
#include <pwd.h>
#include <fcntl.h>
#include <pthread.h>
#include <semaphore.h>
#include <freerdp/freerdp.h>
#include <freerdp/chanman.h>
#include <freerdp/kbd.h>
#include "xf_types.h"
#include "xf_win.h"
#include "xf_keyboard.h"

#ifdef __APPLE__
#include <mach/mach.h>
#include <mach/semaphore.h>
#include <mach/task.h>
#endif

#define MAX_PLUGIN_DATA 20

static sem_t g_sem;
static volatile int g_thread_count = 0;

void freerdp_sem_create(void * sem_struct, int iv)
{
#ifdef __APPLE__
	semaphore_create(mach_task_self(), (semaphore_t *)sem_struct, SYNC_POLICY_FIFO, iv);
#else
	int pshared = 0;
	sem_init((sem_t *)sem_struct, pshared, iv);
#endif
}

void freerdp_sem_signal(void * sem_struct)
{
#ifdef __APPLE__
	semaphore_signal(*((semaphore_t *)sem_struct));
#else
	sem_post((sem_t *)sem_struct);
#endif
}

void freerdp_sem_wait(void * sem_struct)
{
#ifdef __APPLE__
	semaphore_wait(*((semaphore_t *)sem_struct));
#else
	sem_wait((sem_t *)sem_struct);
#endif
}

static int
set_default_params(xfInfo * xfi)
{
	rdpSet * settings;

	settings = xfi->settings;
	memset(settings, 0, sizeof(rdpSet));
	gethostname(settings->hostname, sizeof(settings->hostname) - 1);
	settings->width = 1024;
	settings->height = 768;
	strcpy(settings->server, "127.0.0.1");
	strcpy(settings->username, "guest");
	settings->tcp_port_rdp = 3389;
	settings->encryption = 1;
	settings->server_depth = 16;
	settings->bitmap_cache = 1;
	settings->bitmap_compression = 1;
	settings->desktop_save = 0;
	settings->performanceflags =
		PERF_DISABLE_WALLPAPER | PERF_DISABLE_FULLWINDOWDRAG | PERF_DISABLE_MENUANIMATIONS;
	settings->off_screen_bitmaps = 1;
	settings->triblt = 0;
	settings->new_cursors = 1;
	settings->rdp_version = 5;
	settings->rdp_security = 1;
#ifndef DISABLE_TLS
	settings->tls_security = 1;
	settings->nla_security = 1;
#endif
	xfi->fullscreen = xfi->fs_toggle = 0;
	xfi->decoration = 1;
	return 0;
}

static int
out_args(void)
{
	char help[] =
		"\n"
		"FreeRDP - A Free Remote Desktop Protocol Client\n"
		"See http://freerdp.sourceforge.net for more information\n"
		"\n"
		"Usage: xfreerdp [options] server:port\n"
		"\t-a: color depth (8, 15, 16, 24 or 32)\n"
		"\t-u: username\n"
		"\t-p: password\n"
		"\t-d: domain\n"
		"\t-k: keyboard layout ID\n"
		"\t--kbd-list: list all keyboard layout IDs\n"
		"\t-s: shell\n"
		"\t-c: directory\n"
		"\t-g: geometry, using format WxH or X%, default is 1024x768\n"
		"\t-t: alternative port number, default is 3389\n"
		"\t-n: hostname\n"
		"\t-o: console audio\n"
		"\t-0: console session\n"
		"\t-f: fullscreen mode\n"
		"\t-D: hide window decorations\n"
		"\t-z: enable bulk compression\n"
		"\t-x: performance flags (m, b or l for modem, broadband or lan)\n"
		"\t-X: embed into another window with a given XID.\n"
#ifndef DISABLE_TLS
		"\t--no-rdp: disable Standard RDP encryption\n"
		"\t--no-tls: disable TLS encryption\n"
		"\t--no-nla: disable network level authentication\n"
		"\t--sec: force protocol security (rdp, tls or nla)\n"
#endif
		"\t--plugin: load a virtual channel plugin\n"
		"\t--no-osb: disable off screen bitmaps, default on\n"
		"\t--version: Print out the version and exit\n"
		"\t-h: show this help\n";
	printf("%s\n", help);
	return 0;
}

/* Returns "true" on errors or other reasons to not continue normal operation */
static int
process_params(xfInfo * xfi, int argc, char ** argv, int * pindex)
{
	rdpSet * settings;
	rdpKeyboardLayout * layouts;
	char * p;
	RD_PLUGIN_DATA plugin_data[MAX_PLUGIN_DATA + 1];
	int index;
	int i, j;
	struct passwd * pw;
	int num_extensions;

	set_default_params(xfi);
	settings = xfi->settings;
	num_extensions = 0;
	p = getlogin();
	i = sizeof(settings->username) - 1;
	if (p != 0)
	{
		strncpy(settings->username, p, i);
	}
	else
	{
		pw = getpwuid(getuid());
		if (pw != 0)
		{
			if (pw->pw_name != 0)
			{
				strncpy(settings->username, pw->pw_name, i);
			}
		}
	}

	if (argc < *pindex + 1)
	{
		if (*pindex == 1)
			printf("no parameters specified\n");
		return 1;
	}
	while (*pindex < argc)
	{
		if (strcmp("-a", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing server depth\n");
				return 1;
			}
			settings->server_depth = atoi(argv[*pindex]);
		}
		else if (strcmp("-u", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing username\n");
				return 1;
			}
			strncpy(settings->username, argv[*pindex], sizeof(settings->username) - 1);
			settings->username[sizeof(settings->username) - 1] = 0;
		}
		else if (strcmp("-p", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing password\n");
				return 1;
			}
			strncpy(settings->password, argv[*pindex], sizeof(settings->password) - 1);
			settings->password[sizeof(settings->password) - 1] = 0;
			settings->autologin = 1;

			/*
			 * Overwrite original password which could be revealed by a simple "ps aux" command.
			 * This approach won't hide the password length, but it is better than nothing.
			 */
			
			memset(argv[*pindex], '*', strlen(argv[*pindex]));
		}
		else if (strcmp("-d", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing domain\n");
				return 1;
			}
			strncpy(settings->domain, argv[*pindex], sizeof(settings->domain) - 1);
			settings->domain[sizeof(settings->domain) - 1] = 0;
		}
		else if (strcmp("-k", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing keyboard layout ID\n");
				return 1;
			}
			sscanf(argv[*pindex], "%X", &(xfi->keyboard_layout_id));
			printf("keyboard layout ID: %X\n", xfi->keyboard_layout_id);
		}
		else if (strcmp("--kbd-list", argv[*pindex]) == 0)
		{
			layouts = freerdp_kbd_get_layouts(RDP_KEYBOARD_LAYOUT_TYPE_STANDARD);
			printf("\nKeyboard Layouts\n");
			for (i = 0; layouts[i].code; i++)
				printf("0x%08X\t%s\n", layouts[i].code, layouts[i].name);
			free(layouts);

			layouts = freerdp_kbd_get_layouts(RDP_KEYBOARD_LAYOUT_TYPE_VARIANT);
			printf("\nKeyboard Layout Variants\n");
			for (i = 0; layouts[i].code; i++)
				printf("0x%08X\t%s\n", layouts[i].code, layouts[i].name);
			free(layouts);

			layouts = freerdp_kbd_get_layouts(RDP_KEYBOARD_LAYOUT_TYPE_IME);
			printf("\nKeyboard Input Method Editors (IMEs)\n");
			for (i = 0; layouts[i].code; i++)
				printf("0x%08X\t%s\n", layouts[i].code, layouts[i].name);
			free(layouts);

			return 1;
		}
		else if (strcmp("-s", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing shell\n");
				return 1;
			}
			strncpy(settings->shell, argv[*pindex], sizeof(settings->shell) - 1);
			settings->shell[sizeof(settings->shell) - 1] = 0;
		}
		else if (strcmp("-c", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing directory\n");
				return 1;
			}
			strncpy(settings->directory, argv[*pindex], sizeof(settings->directory) - 1);
			settings->directory[sizeof(settings->directory) - 1] = 0;
		}
		else if (strcmp("-g", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing width\n");
				return 1;
			}
			settings->width = strtol(argv[*pindex], &p, 10);
			if (*p == 'x')
			{
				settings->height = strtol(p + 1, &p, 10);
			}
			if (*p == '%')
			{
				xfi->percentscreen = settings->width;
			}
		}
		else if (strcmp("-t", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing port number\n");
				return 1;
			}
			settings->tcp_port_rdp = atoi(argv[*pindex]);
		}
		else if (strcmp("-n", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing hostname\n");
				return 1;
			}
			strncpy(settings->hostname, argv[*pindex], sizeof(settings->hostname) - 1);
			settings->directory[sizeof(settings->hostname) - 1] = 0;
		}
		else if (strcmp("-o", argv[*pindex]) == 0)
		{
			settings->console_audio = 1;
		}
		else if (strcmp("-0", argv[*pindex]) == 0)
		{
			settings->console_session = 1;
		}
		else if (strcmp("-z", argv[*pindex]) == 0)
		{
			settings->bulk_compression = 1;
		}
		else if (strcmp("--no-osb", argv[*pindex]) == 0)
		{
			settings->off_screen_bitmaps = 0;
		}
		else if (strcmp("-f", argv[*pindex]) == 0)
		{
			xfi->fullscreen = xfi->fs_toggle = 1;
			printf("full screen option\n");
		}
		else if (strcmp("-D", argv[*pindex]) == 0)
		{
			xfi->decoration = 0;
		}
		else if (strcmp("-x", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing performance flag\n");
				return 1;
			}
			if (strncmp("m", argv[*pindex], 1) == 0) /* modem */
			{
				settings->performanceflags = PERF_DISABLE_WALLPAPER |
					PERF_DISABLE_FULLWINDOWDRAG | PERF_DISABLE_MENUANIMATIONS |
					PERF_DISABLE_THEMING;
			}
			else if (strncmp("b", argv[*pindex], 1) == 0) /* broadband */
			{
				settings->performanceflags = PERF_DISABLE_WALLPAPER;
			}
			else if (strncmp("l", argv[*pindex], 1) == 0) /* lan */
			{
				settings->performanceflags = PERF_FLAG_NONE;
			}
			else
			{
				settings->performanceflags = strtol(argv[*pindex], 0, 16);
			}
		} else if (strcmp("-X", argv[*pindex]) == 0) {
			*pindex = *pindex + 1;

			if (*pindex == argc) {
				printf("missing XID\n");
				return 1;
			}

			xfi->embed = strtoul(argv[*pindex], NULL, 16);
			if (!xfi->embed) {
				printf("bad XID\n");
				return 1;
			}
		}
#ifndef DISABLE_TLS
		else if (strcmp("--no-rdp", argv[*pindex]) == 0)
		{
			settings->rdp_security = 0;
		}
		else if (strcmp("--no-tls", argv[*pindex]) == 0)
		{
			settings->tls_security = 0;
		}
		else if (strcmp("--no-nla", argv[*pindex]) == 0)
		{
			settings->nla_security = 0;
		}
		else if (strcmp("--sec", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing protocol security\n");
				return 1;
			}
			if (strncmp("rdp", argv[*pindex], 1) == 0) /* Standard RDP */
			{
				settings->rdp_security = 1;
				settings->tls_security = 0;
				settings->nla_security = 0;
			}
			else if (strncmp("tls", argv[*pindex], 1) == 0) /* TLS */
			{
				settings->rdp_security = 0;
				settings->tls_security = 1;
				settings->nla_security = 0;
			}
			else if (strncmp("nla", argv[*pindex], 1) == 0) /* NLA */
			{
				settings->rdp_security = 0;
				settings->tls_security = 0;
				settings->nla_security = 1;
			}
			else
			{
				printf("unknown protocol security\n");
				return 1;
			}
		}
#endif
		else if (strcmp("--plugin", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing plugin name\n");
				return 1;
			}
			index = *pindex;
			memset(plugin_data, 0, sizeof(plugin_data));
			if (*pindex < argc - 1 && strcmp("--data", argv[*pindex + 1]) == 0)
			{
				*pindex = *pindex + 2;
				i = 0;
				while (*pindex < argc && strcmp("--", argv[*pindex]) != 0 && i < MAX_PLUGIN_DATA)
				{
					plugin_data[i].size = sizeof(RD_PLUGIN_DATA);
					for (j = 0, p = argv[*pindex]; j < 4 && p != NULL; j++)
					{
						plugin_data[i].data[j] = p;
						p = strchr(plugin_data[i].data[j], ':');
						if (p != NULL)
							*p++ = 0;
					}
					*pindex = *pindex + 1;
					i++;
				}
			}
			freerdp_chanman_load_plugin(xfi->chan_man, settings, argv[index], plugin_data);
		}
		else if (strcmp("--ext", argv[*pindex]) == 0)
		{
			*pindex = *pindex + 1;
			if (*pindex == argc)
			{
				printf("missing extension name\n");
				return 1;
			}
			if (num_extensions >= sizeof(settings->extensions) / sizeof(struct rdp_ext_set))
			{
				printf("maximum extensions reached\n");
				return 1;
			}
			index = *pindex;
			snprintf(settings->extensions[num_extensions].name,
				sizeof(settings->extensions[num_extensions].name),
				"%s", argv[index]);
			settings->extensions[num_extensions].data = NULL;
			if (*pindex < argc - 1 && strcmp("--data", argv[*pindex + 1]) == 0)
			{
				*pindex = *pindex + 2;
				settings->extensions[num_extensions].data = argv[*pindex];
				i = 0;
				while (*pindex < argc && strcmp("--", argv[*pindex]) != 0)
				{
					*pindex = *pindex + 1;
					i++;
				}
			}
			num_extensions++;
		}
		else if ((strcmp("-h", argv[*pindex]) == 0) || strcmp("--help", argv[*pindex]) == 0)
		{
			out_args();
			return 1;
		}
		else if (strcmp("--version", argv[*pindex]) == 0)
		{
			printf("This is FreeRDP version %s\n", PACKAGE_VERSION);
			return 1;
		}
		else if (argv[*pindex][0] != '-')
		{
			settings->server[sizeof(settings->server) - 1] = 0;
			if (argv[*pindex][0] == '[' && (p = strchr(argv[*pindex], ']'))
				&& (p[1] == 0 || (p[1] == ':' && !strchr(p + 2, ':'))))
			{
				/* Either "[...]" or "[...]:..." with at most one : after the brackets */
				strncpy(settings->server, argv[*pindex] + 1, sizeof(settings->server) - 1);
				if ((p = strchr(settings->server, ']')))
				{
					*p = 0;
					if (p[1] == ':')
						settings->tcp_port_rdp = atoi(p + 2);
				}
			}
			else
			{
				/* Port number is cut off and used if exactly one : in the string */
				strncpy(settings->server, argv[*pindex], sizeof(settings->server) - 1);
				if ((p = strchr(settings->server, ':')) && !strchr(p + 1, ':'))
				{
					*p = 0;
					settings->tcp_port_rdp = atoi(p + 1);
				}
			}
			/* server is the last argument for the current session. arguments
			   followed will be parsed for the next session. */
			*pindex = *pindex + 1;

			return 0;
		}
		else
		{
			printf("invalid option: %s\n", argv[*pindex]);
			return 1;
		}
		*pindex = *pindex + 1;
	}
	printf("missing server name\n");
	return 1;
}

static int
run_xfreerdp(xfInfo * xfi)
{
	rdpInst * inst;
	void * read_fds[32];
	void * write_fds[32];
	int read_count;
	int write_count;
	int index;
	int sck;
	int max_sck;
	fd_set rfds;
	fd_set wfds;

	/* create an instance of the library */
	inst = freerdp_new(xfi->settings);
	if (inst == NULL)
	{
		printf("run_xfreerdp: freerdp_new failed\n");
		return 1;
	}
	if ((inst->version != FREERDP_INTERFACE_VERSION) ||
	    (inst->size != sizeof(rdpInst)))
	{
		printf("run_xfreerdp: freerdp_new size, version / size do not "
		       "match expecting v %d s %d got v %d s %d\n",
		       FREERDP_INTERFACE_VERSION, (int)sizeof(rdpInst),
		       inst->version, inst->size);
		return 1;
	}
	xfi->inst = inst;
	SET_XFI(inst, xfi);
	if (xf_pre_connect(xfi) != 0)
	{
		printf("run_xfreerdp: xf_pre_connect failed\n");
		return 1;
	}
	if (freerdp_chanman_pre_connect(xfi->chan_man, inst) != 0)
	{
		printf("run_xfreerdp: freerdp_chanman_pre_connect failed\n");
		return 1;
	}
	/* call connect */
	printf("keyboard_layout: %X\n", inst->settings->keyboard_layout);
	if (inst->rdp_connect(inst) != 0)
	{
		printf("run_xfreerdp: inst->rdp_connect failed\n");
		return 1;
	}
	if (freerdp_chanman_post_connect(xfi->chan_man, inst) != 0)
	{
		printf("run_xfreerdp: freerdp_chanman_post_connect failed\n");
		return 1;
	}
	if (xf_post_connect(xfi) != 0)
	{
		printf("run_xfreerdp: xf_post_connect failed\n");
		return 1;
	}
	/* program main loop */
	while (1)
	{
		read_count = 0;
		write_count = 0;
		/* get libfreerdp fds */
		if (inst->rdp_get_fds(inst, read_fds, &read_count, write_fds, &write_count) != 0)
		{
			printf("run_xfreerdp: inst->rdp_get_fds failed\n");
			break;
		}
		/* get x fds */
		if (xf_get_fds(xfi, read_fds, &read_count, write_fds, &write_count) != 0)
		{
			printf("run_xfreerdp: xf_get_fds failed\n");
			break;
		}
		/* get channel fds */
		if (freerdp_chanman_get_fds(xfi->chan_man, inst, read_fds, &read_count, write_fds, &write_count) != 0)
		{
			printf("run_xfreerdp: freerdp_chanman_get_fds failed\n");
			break;
		}
		max_sck = 0;
		/* setup read fds */
		FD_ZERO(&rfds);
		for (index = 0; index < read_count; index++)
		{
			sck = (int)(long) (read_fds[index]);
			if (sck > max_sck)
				max_sck = sck;
			FD_SET(sck, &rfds);
		}
		/* setup write fds */
		FD_ZERO(&wfds);
		for (index = 0; index < write_count; index++)
		{
			sck = (int)(long) (write_fds[index]);
			if (sck > max_sck)
				max_sck = sck;
			FD_SET(sck, &wfds);
		}
		/* exit if nothing to do */
		if (max_sck == 0)
		{
			printf("run_xfreerdp: max_sck is zero\n");
			break;
		}
		/* do the wait */
		if (select(max_sck + 1, &rfds, &wfds, NULL, NULL) == -1)
		{
			/* these are not really errors */
			if (!((errno == EAGAIN) ||
				(errno == EWOULDBLOCK) ||
				(errno == EINPROGRESS) ||
				(errno == EINTR))) /* signal occurred */
			{
				printf("run_xfreerdp: select failed\n");
				break;
			}
		}
		/* check the libfreerdp fds */
		if (inst->rdp_check_fds(inst) != 0)
		{
			printf("run_xfreerdp: inst->rdp_check_fds failed\n");
			break;
		}
		/* check x fds */
		if (xf_check_fds(xfi) != 0)
		{
			printf("run_xfreerdp: xf_check_fds failed\n");
			break;
		}
		/* check channel fds */
		if (freerdp_chanman_check_fds(xfi->chan_man, inst) != 0)
		{
			printf("run_xfreerdp: freerdp_chanman_check_fds failed\n");
			break;
		}
	}
	/* cleanup */
	freerdp_chanman_close(xfi->chan_man, inst);
	inst->rdp_disconnect(inst);
	freerdp_free(inst);
	xf_uninit(xfi);
	return 0;
}

static void *
thread_func(void * arg)
{
	xfInfo * xfi;

	xfi = (xfInfo *) arg;
	run_xfreerdp(xfi);
	free(xfi->settings);
	freerdp_chanman_free(xfi->chan_man);
	free(xfi);

	pthread_detach(pthread_self());
	g_thread_count--;
	if (g_thread_count < 1)
	{
		freerdp_sem_signal(&g_sem);
	}
	return NULL;
}

int
main(int argc, char ** argv)
{
	int rv;
	xfInfo * xfi;
	pthread_t thread;
	int index = 1;

	setlocale(LC_CTYPE, "");
	if (argc == 1)
	{
		out_args();
		return 0;
	}
	if (!freerdp_global_init())
	{
		printf("Error initializing freerdp\n");
		return 1;
	}
	freerdp_chanman_init();

	freerdp_sem_create(&g_sem, 0);
	
	while (1)
	{
		xfi = (xfInfo *) malloc(sizeof(xfInfo));
		memset(xfi, 0, sizeof(xfInfo));
		xfi->settings = (rdpSet *) malloc(sizeof(rdpSet));
		xfi->chan_man = freerdp_chanman_new();
		rv = process_params(xfi, argc, argv, &index);
		if (rv)
		{
			free(xfi->settings);
			freerdp_chanman_free(xfi->chan_man);
			free(xfi);
			break;
		}

		xf_kb_init(xfi->keyboard_layout_id);
		printf("starting thread %d to %s:%d\n", g_thread_count,
			xfi->settings->server, xfi->settings->tcp_port_rdp);
		if (pthread_create(&thread, 0, thread_func, xfi) == 0)
		{
			g_thread_count++;
		}
	}

	if (g_thread_count > 0)
	{
		printf("main thread, waiting for all threads to exit\n");
		freerdp_sem_wait(&g_sem);
		printf("main thread, all threads did exit\n");
	}

	freerdp_chanman_uninit();
	freerdp_global_finish();
	return 0;
}
